tt_metal/tt_stl/any_range.hpp

```diff
@@ -0,0 +1,831 @@
+        return iterator_adaptor()++;
+    }
+
+    // bidirectional methods
+   private:
+    struct enable_if_bidirectional
+        : detail::enable_if_base_of<std::bidirectional_iterator_tag, iterator_category, enable_if_bidirectional> {};
+
+   public:
+    template <class TEnable = enable_if_bidirectional, class = typename TEnable::type>
+    AnyIterator &operator--() {
+        static_assert(std::is_same_v<TEnable, enable_if_bidirectional>);
+        --iterator_adaptor();
+        return *this;
+    }
+
+    template <class TEnable = enable_if_bidirectional, class = typename TEnable::type>
+    [[nodiscard]] AnyIterator operator--(int) {
+        static_assert(std::is_same_v<TEnable, enable_if_bidirectional>);
+        return iterator_adaptor()--;
+    }
+
+    // random access methods
+   private:
+    struct enable_if_random_access
+        : detail::enable_if_base_of<std::random_access_iterator_tag, iterator_category, enable_if_random_access> {};
+
+   public:
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    AnyIterator &operator+=(difference_type offset) {
+        static_assert(std::is_same_v<TEnable, enable_if_random_access>);
+        iterator_adaptor() += offset;
+        return *this;
+    }
+
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    AnyIterator &operator-=(difference_type offset) {
+        static_assert(std::is_same_v<TEnable, enable_if_random_access>);
+        iterator_adaptor() += -offset;
+        return *this;
+    }
+
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    [[nodiscard]] AnyIterator operator+(difference_type offset) const {
+        static_assert(std::is_same_v<TEnable, enable_if_random_access>);
+        return iterator_adaptor() + offset;
+    }
+
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    [[nodiscard]] friend AnyIterator operator+(difference_type offset, const AnyIterator &other) {
+        static_assert(std::is_same_v<TEnable, enable_if_random_access>);
+        return other.iterator_adaptor() + offset;
+    }
+
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    [[nodiscard]] AnyIterator operator-(difference_type offset) const {
+        static_assert(std::is_same_v<TEnable, enable_if_random_access>);
+        return iterator_adaptor() + -offset;
+    }
+
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    [[nodiscard]] difference_type operator-(const AnyIterator &other) const {
+        static_assert(std::is_same_v<TEnable, enable_if_random_access>);
+        return iterator_adaptor() - other.iterator_adaptor();
+    }
+
+    template <class TEnable = enable_if_random_access, class = typename TEnable::type>
+    [[nodiscard]] reference operator[](difference_type offset) const {
```

Comment:
I think you need to consider adding noexcept(noexcept(iterator adaptor call)) in most of the functions.

Reply:
It can't be noexcept, even conditionally, unless every possible type being erased was noexcept for the corresponding call, so there's no point in adding that.